### DISCUSSION TITLE
autorun() without disposal doesn't seem to memory-leak?

### DISCUSSION THREAD

**Discussion by brundonsmith** [mobxjs/mobx#3185] [Q&A] [answered]
I'm trying to get a deeper understanding of how MobX works, particularly around reactions and disposal of reactions. To test things out, I wrote the following code:
```javascript
import * as mobx from "https://jspm.dev/npm:mobx@6.3.3"

function junk() {
    let arr = []
    for (let i = 0; i < 10000000; i++) {
        arr.push(Math.random())
    }
    return arr
}

console.log('running...')
for (let i = 0; i < 10; i++) {
    const j = mobx.observable.box(junk(), { deep: false });
    mobx.autorun(() => {
        console.log(j.get().length)
    })
}
console.log('done')
```
The `junk()` function generates a large array of noise data, so that its allocation and freeing will be clearly visible in the profiler. The theory, in this code, is that the `junk()` result created in each iteration of this loop will not be garbage-collected because the `autorun()` is referencing it (and is not disposed). But here's what I see in the profiler (the blue line is heap memory):

![image](https://user-images.githubusercontent.com/3674979/141601928-bbf045df-277a-41bc-b8db-a9125a15d446.png)

It very much looks like the arrays are getting collected. Is there a problem with my test code, or a gap in my mental model?

----------

Edit: Follow-up question; the original thing I was trying to understand was for reaction(), whether disposals are necessary to free values referenced in the data function, or the effect function, or both. But the above case makes that question moot, until it's resolved. But any further insight would be appreciated!

One possibility: does MobX have a way of knowing when a reaction is "no longer being used" in some cases (similar to what it does for @computed), allowing it to automatically dispose them in certain situations?

**Comment by mweststrate** [comment, marked as answer]
Both junk and autorun run out of scope, so they can be gc-ed together. The
problem arises when a reference is kept to one of those. Typically a single
observable that is used by the autoruns is enough for that.

**Reply by brundonsmith** [reply]
Hmm, ok. I was thinking the reaction gets pinned to some global location where it can't be freed (and holds its references to everything it references) until it's explicitly told to free itself (releasing everything else it references).

So if no relevant "part" of the reaction is still in memory, it will be freed?

I just tested the following scenario and it _looks_ like it does indeed leak. Is this the kind of thing you had in mind?
```javascript
import * as mobx from "https://jspm.dev/npm:mobx@6.3.3"

function junk() {
    let arr = []
    for (let i = 0; i < 10000000; i++) {
        arr.push(Math.random())
    }
    return arr
}

console.log('running...')
const obs = mobx.observable({ val: 0 })
for (let i = 0; i < 10; i++) {
    const j = mobx.observable.box(junk(), { deep: false });
    mobx.reaction(() => obs.val, () => {
        console.log(j.get().length)
    })
}
console.log('done')
```

**Reply by brundonsmith** [reply]
One more angle (thank you for taking the time to answer these questions!): my end goal here is to figure out the circumstances under which it's safe to *not* dispose reactions (`reaction()` and/or `autorun()`). If there's a simple way to define that set of cases, I would be very grateful/excited to learn it!

**Comment by mweststrate** [comment]
That was indeed what I had in mind. Since it can be very subtle, just
always dispose anything that doesnt span the entire lifecycle if the app.
